blockonomics-woocommerce.php

```diff
@@ -401,14 +401,62 @@ function bnomics_async_scripts($url)
 register_activation_hook( __FILE__, 'blockonomics_activation_hook' );
 add_action('admin_notices', 'blockonomics_plugin_activation');
 
+global $blockonomics_db_version;
+$blockonomics_db_version = '1.0';
+
 function blockonomics_activation_hook() {
     if(!is_plugin_active('woocommerce/woocommerce.php'))
     {
         trigger_error(__( 'Wordpress Bitcoin Payments - Blockonomics requires WooCommerce plugin to be installed and active.', 'blockonomics-bitcoin-payments' ).'<br>', E_USER_ERROR);
     }
+
+    // Create blockonomics_orders table
+    // https://codex.wordpress.org/Creating_Tables_with_Plugins
+    global $wpdb;
+    global $blockonomics_db_version;
+
+    $installed_ver = get_option( "blockonomics_db_version" );
+
+    if ( $installed_ver != $blockonomics_db_version ) {
+        $table_name = $wpdb->prefix . 'blockonomics_orders';
+        $charset_collate = $wpdb->get_charset_collate();
+        $sql = "CREATE TABLE IF NOT EXISTS $table_name (
+            order_id int NOT NULL,
+            status int NOT NULL,
+            crypto varchar(3) NOT NULL,
+            address varchar(191) NOT NULL,
+            timestamp int,
+            time_remaining int,
+            satoshi int,
+            currency varchar(3),
+            value longtext,
+            txid text,
+            UNIQUE (address),
+            PRIMARY KEY  (address)
+        ) $charset_collate;";
+        require_once( ABSPATH . 'wp-admin/includes/upgrade.php' );
+        dbDelta( $sql );
+
+        // Example function to demonstrate table changes between upgrade versions
+        // if ($installed_ver < 1.0) {
+        //     $wpdb->query("ALTER TABLE $table_name DROP transaction;");
+        // }
+
+        update_option( 'blockonomics_db_version', $blockonomics_db_version );
+    }
+
     set_transient( 'blockonomics_activation_hook_transient', true, 5);
 }
 
+// Since WP 3.1 the activation function registered with register_activation_hook() is not called when a plugin is updated.
+function blockonomics_update_db_check() {
+    global $blockonomics_db_version;
+    if ( get_site_option( 'blockonomics_db_version' ) != $blockonomics_db_version ) {
+        blockonomics_activation_hook();
+    }
+}
+add_action( 'plugins_loaded', 'blockonomics_update_db_check' );
```

Comment:
Since we are anyway using the *plugins_loaded* hook, I think it is uncessary to add the update code inside activation hook. This will call it to be invoked twice - once on activation and another on *plugins_loaded* . We can move this code to a seperate function

Reply:
WordPress recommends the upgrade be structured in this manner. https://codex.wordpress.org/Creating_Tables_with_Plugins#Adding_an_Upgrade_Function

I have separated the install and upgrade function in ef135d6. Users upgrading from a previous version will now be required to deactivate and activate the plugin in order to create the table.

Reply:
Fixed logic for existing users in c79436f